File: python/epi/interconverting_string_integer.py
#!/usr/bin/env python3
#
# Description:
#
# Interconvert string and integer. Handle negative values.

from rlib.testing_utils import expect

def int_to_string(value):
    if value == 0: return '0'
    if value < 0:
        neg = True
        value = -value
    else:
        neg = False
    res = []
    while value > 0:
        res.append(str(value % 10))
        value //= 10
    if neg: res.append('-')
    res.reverse()
    return ''.join(res)

def string_to_int(value):
    res = 0
    for s in value[value[0] == '-':]:
        res *= 10
        res += ord(s) - ord('0')
    return -res if value[0] == '-' else res

def small_test():
    expect(int_to_string(0)).to_be('0')
    expect(int_to_string(-1)).to_be('-1')
    expect(int_to_string(1)).to_be('1')
    expect(int_to_string(1000)).to_be('1000')
    expect(int_to_string(2147483646)).to_be('2147483646')
    expect(int_to_string(-2147483646)).to_be('-2147483646')
    expect(string_to_int('0')).to_be(0)
    expect(string_to_int('-1')).to_be(-1)
    expect(string_to_int('1')).to_be(1)
    expect(string_to_int('1000')).to_be(1000)
    expect(string_to_int('2147483646')).to_be(2147483646)
    expect(string_to_int('-2147483646')).to_be(-2147483646)

if __name__ == '__main__':
    small_test()

File: java/epi_old/InterconvertingStringInteger.java
// Description:
//
// Convert string to int and int to string.

package epi;

import java.lang.StringBuilder;
import static rlib.TestingUtils.expect;

public class InterconvertingStringInteger {

  public static String intToString(int x) {
    boolean isNeg = false;
    if (x < 0) {
      isNeg = true;
      x = -x;
    }
    StringBuilder sb = new StringBuilder();
    if (x == 0) {
      sb.append('0');
    } else {
      while (x > 0) {
        sb.append((char)('0' + x % 10));
        x /= 10;
      }
    }
    if (isNeg) sb.append('-');
    return sb.reverse().toString();
  }

  private static int stringToInt(String s) {
    boolean isNeg = (s.charAt(0) == '-') ? true : false;
    int res = 0;
    for (int i = (isNeg) ? 1 : 0; i < s.length(); i++) {
      res = (res * 10) + s.charAt(i) - '0';
    }
    return (isNeg) ? -res : res;
  }

  private static void smallTest() {
    expect(intToString(0)).toBe("0");
    expect(intToString(-1)).toBe("-1");
    expect(intToString(1)).toBe("1");
    expect(intToString(2)).toBe("2");
    expect(intToString(-2)).toBe("-2");
    expect(intToString(9)).toBe("9");
    expect(intToString(10)).toBe("10");
    expect(intToString(123)).toBe("123");
    expect(intToString(Integer.MAX_VALUE)).toBe(Integer.toString(Integer.MAX_VALUE));

    expect(stringToInt("0")).toBe(0);
    expect(stringToInt("-1")).toBe(-1);
    expect(stringToInt("1")).toBe(1);
    expect(stringToInt("2")).toBe(2);
    expect(stringToInt("-2")).toBe(-2);
    expect(stringToInt("9")).toBe(9);
    expect(stringToInt("10")).toBe(10);
    expect(stringToInt("123")).toBe(123);
    expect(stringToInt(Integer.toString(Integer.MAX_VALUE))).toBe(Integer.MAX_VALUE);
  }

  public static void main(String[] args) {
    smallTest();
  }

}